### ISSUE TITLE
Sparse matrix accessor is much slower with newer versions of g++

### ISSUE THREAD

**Issue by privefl** [RcppCore/RcppArmadillo#234]
I have a function that stores results in a sparse matrix. After upgrading from g++ v4.8.x to v5.4.0, this function has been 10 times slower (which is not the case anymore if I use a dense matrix).

A minimal example that tries to reproduce what my function does:
```
#include <RcppArmadillo.h>
using namespace Rcpp;

// [[Rcpp::export]]
void test_acc(int n, int p, int K) {

  NumericMatrix X(n, p);
  for (int j = 0; j < p; j++)
    for (int k = 0; k < K; k++)
      X(0, j) = norm_rand();
}

// [[Rcpp::export]]
void test_sp_acc(int n, int p, int K) {

  arma::sp_mat X(n, p);
  for (int j = 0; j < p; j++)
    for (int k = 0; k < K; k++)
      X(0, j) = norm_rand();
}


/*** R
microbenchmark::microbenchmark(
  test_acc(1e3, 2e3, 200),
  test_sp_acc(1e3, 2e3, 200)
)
### RcppArmadillo v0.8.500.0 & g++ v5.4.0
# Unit: milliseconds
#                         expr      min       lq     mean   median       uq       max neval
#    test_acc(1000, 2000, 200) 16.43358 17.02553 19.13284 17.81334 18.15556 121.57572   100
# test_sp_acc(1000, 2000, 200) 24.13308 24.80901 25.53917 24.98361 25.24921  53.59381   100
*/
```

I don't know how to test it with an older version of g++ though.

**Comment by dselivanov** [comment]
Not clear what are you trying to benchmark - matrix creation or access time. I suggest you to consider make more isolation in the benchmark. 
Also could you add timings for g++ 4.8?

**Comment by privefl** [comment]
As I said, I don't know how to test it with an older version of g++.
If you know how to do that (using Makevars maybe?), please tell me so that I can improve this issue.

**Comment by dselivanov** [comment]
Specify compiler in `~/.R/Makevars`:
```
CXX=/path/to/g++-4.8
CXX1X=/path/to/g++-4.8
CXX11=/path/to/g++-4.8
```

**Comment by privefl** [comment]
I get the following compilation errors:
```
g++-4.8: error: unrecognized command line option ‘-fstack-protector-strong’
g++-4.8: error: unrecognized command line option ‘-Wdate-time’
```

**Comment by eddelbuettel** [comment]
There is nothing operational here so I will close in due course.

If you _really_ suspect a compiler regression nobody else notice you _could_ start eg with plain Docker containers of given Ubuntu vintages, matbe 14.04 or 16.04 (guessing here for the versions) and/or use the PPA with compiler nightlies to pick particular versions. I use that PPA eg [here](https://github.com/eddelbuettel/rcpptoml/blob/master/.travis.yml#L14).

**Comment by privefl** [comment]
Sorry, but I have no idea what a PPA is. And I've never used docker before.

Can anyone test this code with g++ 4.8?

**Comment by dselivanov** [comment]
@privefl I suggest you to first benchmark your code with different versions of `RcppArmadillo`. There were significant changes related to sparse matrices between 7* and 8* series.

**Comment by eddelbuettel** [comment]
This whole issue appears to be a wee bit premature, short of actual evidence and not showing any sign of pending improvement -- so closing.

**Comment by conradsnicta** [comment]
@privefl - This issue is fixed in Armadillo 8.600 and the corresponding RcppArmadillo release.  If not, please file a bug report at the Armadillo repo at GitLab: https://gitlab.com/conradsnicta/armadillo-code

**Comment by privefl** [comment]
This fixed the problem.
Thanks @conradsnicta and @dselivanov for the help.
